src/Symfony/Component/Security/Http/RememberMe/PersistentRememberMeHandler.php

```diff
@@ -0,0 +1,110 @@
+use Symfony\Component\Security\Core\Exception\AuthenticationException;
+use Symfony\Component\Security\Core\Exception\CookieTheftException;
+use Symfony\Component\Security\Core\User\UserInterface;
+use Symfony\Component\Security\Core\User\UserProviderInterface;
+
+/**
+ * Implements remember me tokens using a {@see TokenProviderInterface}.
+ *
+ * This requires storing remember me tokens in a database. This allows
+ * more control over the invalidation of remember me tokens. See
+ * {@see SignatureRememberMeHandler} if you don't want to use a database.
+ *
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ *
+ * @experimental in 5.3
+ */
+final class PersistentRememberMeHandler extends AbstractRememberMeHandler
+{
+    private $tokenProvider;
+    private $secret;
+
+    public function __construct(TokenProviderInterface $tokenProvider, string $secret, UserProviderInterface $userProvider, RequestStack $requestStack, array $options, ?LoggerInterface $logger = null)
+    {
+        parent::__construct($userProvider, $requestStack, $options, $logger);
+
+        $this->tokenProvider = $tokenProvider;
+        $this->secret = $secret;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function createRememberMeCookie(UserInterface $user): void
+    {
+        $series = base64_encode(random_bytes(64));
+        $tokenValue = $this->generateHash(base64_encode(random_bytes(64)));
+        $token = new PersistentToken(\get_class($user), $user->getUsername(), $series, $tokenValue, new \DateTime());
+
+        $this->tokenProvider->createNewToken($token);
+        $this->createCookie(RememberMeDetails::fromPersistentToken($token, time() + $this->options['lifetime']));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function processRememberMe(RememberMeDetails $rememberMeDetails, UserInterface $user): void
+    {
+        [$series, $tokenValue] = explode(':', $rememberMeDetails->getValue());
+        $persistentToken = $this->tokenProvider->loadTokenBySeries($series);
+        if (!hash_equals($persistentToken->getTokenValue(), $tokenValue)) {
+            throw new CookieTheftException('This token was already used. The account is possibly compromised.');
+        }
+
+        if ($persistentToken->getLastUsed()->getTimestamp() + $this->options['lifetime'] < time()) {
+            throw new AuthenticationException('The cookie has expired.');
+        }
+
+        $tokenValue = base64_encode(random_bytes(64));
+        $this->tokenProvider->updateToken($series, $this->generateHash($tokenValue), new \DateTime());
+
+        $this->createCookie($rememberMeDetails->withValue($tokenValue));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function clearRememberMeCookie(): void
+    {
+        parent::clearRememberMeCookie();
+
+        $cookie = $this->requestStack->getMasterRequest()->cookies->get($this->options['name']);
+        if (null === $cookie) {
```

Comment:
What is this check for? If the request doesn't have the cookie, we *avoid* deleting it from the storage?

Reply:
We need the cookie of the request to know which one to delete from the storage. In theory, `ParameterBag::get()` may return null, so that's what this check is taking care of.